### PR TITLE
Add more docs to codegen

### PR DESCRIPTION
**Pull request by rcoh**
## Motivation and Context
@crisidev pointed out that codegen is woefully under documented. This is really only the tip of the iceberg, but we need to start somewhere :-)

### CODE REVIEW

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/CodegenVisitor.kt [hunk 1]
```diff
@@ -73,19 +76,40 @@ class CodegenVisitor(context: PluginContext, private val codegenDecorator: RustC
...
+     * - NormalizeOperations by ensuring every operation has an input & output shape
+     * - Drop unsupported event stream operations from the model
+     * - Normalize event stream operations
+     */
     private fun baselineTransform(model: Model) =
         model.let(RecursiveShapeBoxer::transform)
             .letIf(settings.codegenConfig.addMessageToErrors, AddErrorMessage::transform)
             .let(OperationNormalizer::transform)
             .let { RemoveEventStreamOperations.transform(it, settings) }
             .let(EventStreamNormalizer::transform)
```
Comment (jdisanti) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/CodegenVisitor.kt:92]: I think these will be easier to maintain if we inline them over each `.let(...)`

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/protocol/ProtocolGenerator.kt [hunk 2]
```diff
@@ -26,14 +27,44 @@ import software.amazon.smithy.rust.codegen.smithy.generators.operationBuildError
 import software.amazon.smithy.rust.codegen.smithy.protocols.Protocol
 import software.amazon.smithy.rust.codegen.util.inputShape
 
+/**
+ * Request Body Generator
+ *
+ * **Note:** There is only one real implementation of this trait. The other implementation is test only.
```
Comment (jdisanti) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/protocol/ProtocolGenerator.kt:33]:
```suggestion
 * **Note:** There is only one real implementation of this interface. The other implementation is test only.
```
😉

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/protocol/ProtocolGenerator.kt [hunk 3]
```diff
@@ -26,14 +27,44 @@ import software.amazon.smithy.rust.codegen.smithy.generators.operationBuildError
...
+     *
+     * This should be an expression that returns an `SdkBody`
+     */
     fun generateBody(writer: RustWriter, self: String, operationShape: OperationShape)
 }
 
+/**
+ * Protocol Trait implementation generator
+ *
+ * **Note:** There is only one real implementation of this trait. The other implementation is test only.
```
Comment (jdisanti) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/protocol/ProtocolGenerator.kt:60]:
```suggestion
 * **Note:** There is only one real implementation of this interface. The other implementation is test only.
```

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/protocol/ProtocolGenerator.kt [hunk 4]
```diff
@@ -43,8 +74,19 @@ interface ProtocolTraitImplGenerator {
  */
 open class ProtocolGenerator(
     codegenContext: CodegenContext,
+    /**
+     * `Protocol` contains all protocol specific information. Each smithy protocol, eg. RestJson, RestXml, etc. will
+     * have their own implementation of the protocol interface which defines how an input shape becomes and http::Request
+     * and an output shape is build from an http::Response.
+     */
     private val protocol: Protocol,
+    /** Operations generate a `make_operation(&config)` method to build a `smithy_http::Operation` that can be dispatched
```
Comment (jdisanti) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/protocol/ProtocolGenerator.kt:83]: Nit: generally, the `/**` has the entire line to itself, but not a big deal if ktlint doesn't care.

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/protocol/ProtocolGenerator.kt [hunk 5]
```diff
@@ -63,6 +105,13 @@ open class ProtocolGenerator(
         "operation" to RuntimeType.operationModule(runtimeConfig),
     )
 
+    /**
+     * Render all code required for serializing requests and deserializing responses for the operation
+     *
+     * This primarily relies on two components: :
+     * 1. [traitGenerator]: Generate implementations of the `ParseHttpResponse` trait for the operations
+     * 2. [bod]
```
Comment (jdisanti) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/protocol/ProtocolGenerator.kt:113]: What is bod?

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/protocol/ProtocolGenerator.kt [hunk 6]
```diff
@@ -63,6 +105,13 @@ open class ProtocolGenerator(
         "operation" to RuntimeType.operationModule(runtimeConfig),
     )
 
+    /**
+     * Render all code required for serializing requests and deserializing responses for the operation
+     *
+     * This primarily relies on two components: :
```
Comment (jdisanti) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/protocol/ProtocolGenerator.kt:111]:
```suggestion
     * This primarily relies on two components:
```

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/protocol/ProtocolGenerator.kt (hunk 5 above)
Comment (rcoh) [reply codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/protocol/ProtocolGenerator.kt]: good catch, fixed

**Comment by crisidev** [comment]
Wow @rcoh, this is amazing! Thanks a million for doing this. I'll review the docs tomorrow mornign my time !

**Pull request merged**